packages/data/src/registry.js

```diff
@@ -180,19 +198,34 @@ export function createRegistry( storeConfigs = {} ) {
 				let fulfillment = resolver.fulfill( state, ...args );
 
 				// Attempt to normalize fulfillment as async iterable.
-				fulfillment = toAsyncIterable( fulfillment );
-				if ( ! isAsyncIterable( fulfillment ) ) {
-					return;
-				}
+				if ( isAsyncIterable( fulfillment ) ) {
+					deprecated( 'Asynchronous iterable resolvers', {
+						alternative: 'the store `controls` object',
+						plugin: 'Gutenberg',
+						version: '3.6',
+					} );
+
+					for await ( const maybeAction of fulfillment ) {
+						// Dispatch if it quacks like an action.
+						if ( isActionLike( maybeAction ) ) {
+							store.dispatch( maybeAction );
+						}
+					}
```

Comment:
Not certain how to do it yet in this PR, but in my other alternative PRs, I extracted all the resolution mechanism (the runtime of the generators) into a `runtime.js` file to handle all the use cases. The result in this file is just something like:

```
startResolution(...)
runtime( fullfillment, () => {
  endResolution()
} )
```

I think it makes this file a bit more clear